### DISCUSSION TITLE
Jest did not exit one second after the test run has completed.

### DISCUSSION THREAD

**Discussion by flextremedev** [TanStack/query#1692] [Q&A]
Hi, I followed the docs testing section and when I run the tests jest logs: 

> Jest did not exit one second after the test run has completed.

This is my code:
```ts
import * as React from 'react';
import {QueryClient, QueryClientProvider, useQuery} from 'react-query';
import {renderHook} from '@testing-library/react-hooks';
export function useCustomHook() {
  const {data} = useQuery('customHook', () => 'Hello');
  return data;
}

const queryClient = new QueryClient();
const wrapper: React.FC = ({children}) => (
  <QueryClientProvider client={queryClient}>{children}</QueryClientProvider>
);
it('should work', () => {
  const {result} = renderHook(() => useCustomHook(), {wrapper});

  expect(result.current).toEqual('Hello');
});
```
Am I missing something or is anyone else experiencing this?
Thanks in advance!

**Comment by TkDodo** [comment]
I put your code in a codesandbox: https://codesandbox.io/s/react-query-basic-testing-nh6tv

what was missing for me was waiting for the query to transition into `success` state, I did what with `waitFor`. Even though you resolve the promise immediately, there will still be one render cycle were things are `loading`

**Reply by flextremedev** [reply]
Yeah, I'm sorry, I forgot to say that this occurs with react-native. I created a repo to reproduce: https://github.com/flextremedev/ReactNativePlayground

Just execute:

> yarn test react-query

Can you confirm this is a bug?

It seems to work fine in react though. 
Still the example should be updated to wait for the "Hello" shouldn't it? You want me to create a PR?

**Reply by TkDodo** [reply]
> Still the example should be updated to wait for the "Hello" shouldn't it? You want me to create a PR?

yes, I think that would be an improvement :)

**Reply by flextremedev** [reply]
Done.
Were you able to reproduce it with react-native?

**Reply by smikheiev** [reply]
@flextremedev @TkDodo So did you guys figure out if it's a bug or not and the way to fix it?

**Reply by TkDodo** [reply]
I'm seeing this:

> Jest did not exit one second after the test run has completed.
>
>This usually means that there are asynchronous operations that weren't stopped in your tests. Consider running Jest with `--detectOpenHandles` to troubleshoot this issue.

and then I do what jest says and run it with `--detectOpenHandles` and then it just works, without any additional output  🤔 

could be related to this jest issue: https://github.com/facebook/jest/issues/9473#issuecomment-579195412

and applying the workaround described there made it work for me:

```diff
- "test": "jest",
+ "test": "jest --detectOpenHandles --forceExit",
```

**Reply by flextremedev** [reply]
that's what I've done. but it's not very satisfying as it's generally bad practice to use forceExit.

**Reply by gabrielbull** [reply]
That's what I do too, but it's not a real solution. Should this discussion be moved to issues?

**Reply by TkDodo** [reply]
Yes, feel free to open an issue. I personally don’t know enough about react native or the internal timers to be of much help here

**Comment by gabrielbull** [comment]
I'm having the same issue

**Comment by dkershner6** [comment]
I am as well.

"@testing-library/react-native": "^7.1.0",
"react-native": "https://github.com/expo/react-native/archive/sdk-38.0.2.tar.gz",
"react-query": "^3.6.0",

**Comment by flextremedev** [comment]
Discussed further in https://github.com/tannerlinsley/react-query/issues/1847
